web_src/js/modules/toast.js

```diff
@@ -0,0 +1,67 @@
+import {htmlEscape} from 'escape-goat';
+import {svg} from '../svg.js';
+
+const levels = {
+  info: {
+    icon: 'octicon-check',
+    background: 'var(--color-green)',
+    duration: 2000,
+  },
+  warning: {
+    icon: 'gitea-exclamation',
+    background: 'var(--color-orange)',
+    duration: -1, // needs to be clicked away
+  },
+  error: {
+    icon: 'gitea-exclamation',
+    background: 'var(--color-red)',
+    duration: -1, // needs to be clicked away
+  },
+};
+
+// See https://github.com/apvarun/toastify-js#api for options
+async function showToast(message, level, {gravity, position, duration, ...other} = {}) {
+  if (!message) return;
+
+  const {default: Toastify} = await import(/* webpackChunkName: 'toastify' */'toastify-js');
+  const {icon, background, duration: levelDuration} = levels[level ?? 'info'];
+
+  const toast = Toastify({
+    text: `
+      <div class='toast-icon'>${svg(icon)}</div>
+      <div class='toast-body'>${htmlEscape(message)}</div>
+      <button class='toast-close'>${svg('octicon-x')}</button>
+    `,
+    escapeMarkup: false,
+    gravity: gravity ?? 'top',
+    position: position ?? 'center',
+    duration: duration ?? levelDuration,
+    style: {background},
+    ...other,
+  });
+
+  toast.showToast();
+
+  toast.toastElement.querySelector('.toast-close').addEventListener('click', () => {
+    toast.removeElement(toast.toastElement);
+  });
+}
+
+export async function showInfo(message, opts) {
+  return await showToast(message, 'info', opts);
+}
+
+export async function showWarning(message, opts) {
+  return await showToast(message, 'warning', opts);
+}
+
+export async function showError(message, opts) {
+  return await showToast(message, 'error', opts);
+}
+
+// export for devtest page in development
+if (process.env.NODE_ENV === 'development') {
+  window.giteaShowInfo = showInfo;
+  window.giteaShowWarning = showWarning;
+  window.giteaShowError = showError;
+}
```

Comment:
Maybe we need a separate JS file for "devtest", to initialize some special elements, including the markdown editor. At the moment, the markdown editor on devtest page can't be initialized either.

Reply:
Needs a separate webpack chunk I guess, but certainly possible.

Reply:
In this PR or following PR? Either is fine to me.

Reply:
Doing it here

Reply:
Done in https://github.com/go-gitea/gitea/pull/25449/commits/d255eac1a382ddb267b42db89a61946026349b5c, now we have a separate chunk for devtest with a JS and CSS file.